### PR TITLE
Support recursive dataclasses

### PR DESCRIPTION
**Pull request by dmontagu**
Closes https://github.com/pydantic/pydantic/issues/5721. This is also a precursor to work on generic dataclasses.

I'll open the PR now, but I think there's a variety of things to establish some consensus on before merging:

* Do we want to unify the `model-not-fully-defined` and (analogous) `dataclass-not-fully-defined` usage errors?
  * Actually, currently in this PR, I have removed `model-not-fully-defined` and replaced it with a common `class-not-fully-defined` error, but would be happy to revert this.
  * If we do revert this change, is `dataclass-not-fully-defined` an acceptable name?
    * Would we essentially fully duplicate the documentation of `model-not-fully-defined` considering it's largely similar and we have a docs section for each usage error code?
* I have renamed `BaseModel.__pydantic_model_complete__` to `BaseModel.__pydantic_complete__` and used the same property on pydantic dataclasses. Any objections?
* With BaseModels, to do an explicit rebuild you call the `BaseModel.model_rebuild` method. I have exported an analogous function called `pydantic.dataclasses.rebuild_dataclass`.
  * Is this name (and import path) acceptable?
  * Also, the function is defined in `pydantic._internal._dataclasses` and re-exported in `pydantic.dataclasses`. Is this acceptable?
  * (Actually, I think Adrian wanted to add a method called `__pydantic_dataclass_rebuild__` or similar to pydantic dataclasses — is that still the case? If so, what should the name of that method be?)

I'll be happy to make any requested changes.

Selected Reviewer: @samuelcolvin

### CODE REVIEW

File: pydantic/_internal/_dataclasses.py
Comment (dmontagu) [inline pydantic/_internal/_dataclasses.py:189]: Need to update this, but wanted to get some review before doing so as I suspect this may be affected by some requested refactors..

File: pydantic/main.py
Comment (dmontagu) [inline pydantic/main.py:702]: Looks like this was just a typo in the old docs

File: tests/test_dataclasses.py
Comment (dmontagu) [inline tests/test_dataclasses.py:1482]: I removed the parts of the original expression here that don't get hit because the error gets raised while evaluating the inner components of the expression.

Note to self: Before merging, should add some checks that are similar in intent to the original expression, but with a deeply nested dict, to make sure the full recursion is working properly.

File: pydantic/dataclasses.py
Comment (dmontagu) [inline pydantic/dataclasses.py:200]: I moved this into the `complete_dataclass` method, so that the fields would also be updated when doing a rebuild.

File: pydantic/_internal/_dataclasses.py
Comment (dmontagu) [inline pydantic/_internal/_dataclasses.py:200]: Not sure if we want to store the original namespace on the dataclasses like we do for BaseModel.

This helps in the scenario where rebuilds are necessary when importing from a separate module and there was a ForwardRef to a type in the module in which the class was _originally_ defined.

This may also be necessary when parametrizing generic cyclically-recursive dataclasses outside the module in which they were defined.

I've left these lines present and commented as a reminder to decide about this. I'm okay either way.

File: docs/usage/errors.md
Comment (samuelcolvin) [inline docs/usage/errors.md:8]:
```suggestion
(such as a subclass of `BaseModel`, or a pydantic `dataclass`) is not defined:
```

File: docs/usage/errors.md
Comment (samuelcolvin) [reply docs/usage/errors.md:71]: `#class-not-fully-defined` LGTM.

File: pydantic/_internal/_dataclasses.py
Comment (samuelcolvin) [inline pydantic/_internal/_dataclasses.py:106]:
```suggestion
        usage_warning_message = (
```

File: pydantic/_internal/_dataclasses.py
Comment (samuelcolvin) [reply pydantic/_internal/_dataclasses.py:200]: I'm pretty neutral / would defer to you on it.

My preference is "simple wins", so I think I'd prefer not to store the namespace, but don't really mind.

File: pydantic/dataclasses.py
Comment (samuelcolvin) [inline pydantic/dataclasses.py:27]:
```suggestion
__all__ = 'dataclass', 'rebuild_dataclass'
```

**Review by samuelcolvin — COMMENTED** [review]
otherwise LGTM.

**Review by samuelcolvin — APPROVED** [review]
conflicts, otherwise LGTM.

Merge when passing.

**Comment by samuelcolvin** [comment]
please update.

File: pydantic/_internal/_dataclasses.py
Comment (dmontagu) [reply pydantic/_internal/_dataclasses.py:216]: Okay, given there is a lot of complexity related to generic dataclasses and how exactly we should handle them, and not handling it properly is not a regression from v1 (as it wasn't handled properly there), my new plan is to not fiddle with the parent namespace, and also not support generic dataclasses in v2 (but perhaps in v2.1, or _maybe_ sooner if we happen to get it all working in time.. but the pressure is off.)

**Pull request merged**
